pkg/controller/virtualmachine/virtualmachine.go

```diff
@@ -0,0 +1,475 @@
+		"name":      vmi.Name,
+		"namespace": vmi.Namespace,
+	}).Debugf("found envMap: %v", envMap)
+	return envMap, nil
+}
+
+func (h *Handler) reconcileVMResourceAllocationAnnotation(vmi *kubevirtv1.VirtualMachineInstance, deviceDetails string) error {
+	logrus.WithFields(logrus.Fields{
+		"name":      vmi.Name,
+		"namespace": vmi.Namespace,
+	}).Debugf("device allocation details: %s", deviceDetails)
+
+	vmObj, err := h.vmCache.Get(vmi.Namespace, vmi.Name)
+	if err != nil {
+		return fmt.Errorf("error fetching vm %s from cache: %v", vmi.Name, err)
+	}
+
+	var currentAnnotationValue string
+	// update device allocation details
+	if vmObj.Annotations == nil {
+		vmObj.Annotations = make(map[string]string)
+	} else {
+		currentAnnotationValue = vmObj.Annotations[v1beta1.DeviceAllocationKey]
+	}
+
+	if currentAnnotationValue != deviceDetails {
+		vmObj.Annotations[v1beta1.DeviceAllocationKey] = deviceDetails
+		_, err = h.vmClient.Update(vmObj)
+	}
+	return err
+}
+
+// OnVMIDeletion will update the VM spec with actual device allocation details
+// this simplifies deletion of devices from the VM object without needing any changes
+// it needs to be done during the VM shutdown avoid the object generation to differ between VM and VMI object
+// and this avoids the generation warning being reported in the UI
+func (h *Handler) OnVMIDeletion(_ string, vmi *kubevirtv1.VirtualMachineInstance) (*kubevirtv1.VirtualMachineInstance, error) {
+	if vmi == nil || vmi.DeletionTimestamp == nil {
+		return vmi, nil
+	}
+
+	// no host or GPU devices present so nothing needed to be done
+	if len(vmi.Spec.Domain.Devices.GPUs) == 0 && len(vmi.Spec.Domain.Devices.HostDevices) == 0 {
+		logrus.WithFields(logrus.Fields{
+			"name":      vmi.Name,
+			"namespace": vmi.Namespace,
+			"hostname":  vmi.Status.NodeName,
+		}).Debug("skipping vmi as it has no hostdevices or GPUs")
+		return vmi, nil
+	}
+
+	vmObj, err := h.vmCache.Get(vmi.Namespace, vmi.Name)
+	if err != nil {
+		return vmi, fmt.Errorf("error fetching vm object for vmi %s/%s: %v", vmi.Namespace, vmi.Name, err)
+	}
+
+	// no annotations so nothing to do
+	if vmObj.Annotations == nil {
+		logrus.WithFields(logrus.Fields{
+			"name":      vmi.Name,
+			"namespace": vmi.Namespace,
+			"hostname":  vmi.Status.NodeName,
+		}).Debug("skipping vmi as it has no annotations defined")
+		return vmi, nil
+	}
+
+	if vmi.Status.NodeName != h.nodeName {
+		logrus.WithFields(logrus.Fields{
+			"name":      vmi.Name,
+			"namespace": vmi.Namespace,
+			"hostname":  vmi.Status.NodeName,
+		}).Debug("skipping vmi as it is not scheduled on current node")
+		return vmi, nil
+	}
+
+	val, ok := vmObj.Annotations[v1beta1.DeviceAllocationKey]
+	if !ok {
+		// no device allocation annotations, nothing to do
+		logrus.WithFields(logrus.Fields{
+			"name":      vmi.Name,
+			"namespace": vmi.Namespace,
+			"hostname":  vmi.Status.NodeName,
+		}).Debug("skipping vmi as it has no device allocation annotation")
+		return vmi, nil
+	}
```

Comment:
NIT: We could remove `if vmObj.Annotations == nil {` this condition to reduce some codes because we only used reading here. Reading a nil map is allowed.